rust/nomad-base/src/agent.rs

```diff
@@ -152,4 +160,35 @@ pub trait NomadAgent: Send + Sync + std::fmt::Debug + AsRef<AgentCore> {
         })
         .instrument(span)
     }
+
+    /// Spawn a task which continuously watch home for getting into failed state
+    /// and resolve once it happened.
+    /// `Reported` flag turns `Ok(())` into `Err(Report)` on failed home.
+    #[allow(clippy::unit_arg)]
+    fn watch_home_fail(&self, interval: u64) -> Instrumented<JoinHandle<Result<()>>> {
+        use nomad_core::Common;
+        let span = info_span!("home_watch");
+        let home = self.home();
+        tokio::spawn(async move {
+            let home = home.clone();
+            loop {
+                if home.state().await? == nomad_core::State::Failed {
+                    return Err(BaseError::FailedHome.into());
+                }
+
+                sleep(Duration::from_secs(interval)).await;
+            }
+        })
+        .instrument(span)
+    }
+
+    /// Returns `true` if home is in failed state. Intended to return once and immediately
+    #[allow(clippy::unit_arg)]
+    fn is_home_failed(&self) -> Instrumented<JoinHandle<Result<bool>>> {
+        use nomad_core::Common;
+        let span = info_span!("check_home_state");
+        let home = self.home();
+        tokio::spawn(async move { Ok(home.state().await? == nomad_core::State::Failed) })
+            .instrument(span)
+    }
```

Comment:
Sorry last change. Since every time we call `is_home_failed` we end up checking if value is true and then return a `FailedHome` error, we can probably just convert this function into the following:
```
fn assert_home_not_failed(&self) -> Result<()> {
    if home.state().await? == nomad_core::State::Failed {
          return Err(BaseError::FailedHome.into());
      }

     Ok(())
}
```

Then instead of calling `is_home_failed`, checking result, then returning error in all the `tokio::spawns`, we can just put this at the top of the task spawn:
```
self.assert_home_not_failed()?;
```

Reply:
good change, though still spawn would need as in some places we cannot really use bare `async fn assert_...` as it moves `self` into the async context, but in `fn run(&self, ...` it is not possible,